pilot/pkg/model/push_status.go

```diff
@@ -196,3 +242,232 @@ func (ps *PushStatus) UpdateMetrics() {
+		}
+		// resolve gateways to bind to
+		for i, g := range rule.Gateways {
+			if g != IstioMeshGateway {
+				rule.Gateways[i] = ResolveShortnameToFQDN(g, r.ConfigMeta).String()
+			}
+		}
+		// resolve host in http route.destination, route.mirror
+		for _, d := range rule.Http {
+			for _, m := range d.Match {
+				for i, g := range m.Gateways {
+					if g != IstioMeshGateway {
+						m.Gateways[i] = ResolveShortnameToFQDN(g, r.ConfigMeta).String()
+					}
+				}
+			}
+			for _, w := range d.Route {
+				w.Destination.Host = ResolveShortnameToFQDN(w.Destination.Host, r.ConfigMeta).String()
+			}
+			if d.Mirror != nil {
+				d.Mirror.Host = ResolveShortnameToFQDN(d.Mirror.Host, r.ConfigMeta).String()
+			}
+		}
+		//resolve host in tcp route.destination
+		for _, d := range rule.Tcp {
+			for _, m := range d.Match {
+				for i, g := range m.Gateways {
+					if g != IstioMeshGateway {
+						m.Gateways[i] = ResolveShortnameToFQDN(g, r.ConfigMeta).String()
+					}
+				}
+			}
+			for _, w := range d.Route {
+				w.Destination.Host = ResolveShortnameToFQDN(w.Destination.Host, r.ConfigMeta).String()
+			}
+		}
+		//resolve host in tls route.destination
+		for _, tls := range rule.Tls {
+			for _, m := range tls.Match {
+				for i, g := range m.Gateways {
+					if g != IstioMeshGateway {
+						m.Gateways[i] = ResolveShortnameToFQDN(g, r.ConfigMeta).String()
+					}
+				}
+			}
+			for _, w := range tls.Route {
+				w.Destination.Host = ResolveShortnameToFQDN(w.Destination.Host, r.ConfigMeta).String()
+			}
+		}
+	}
+
+	return out
+}
+
+// GetServiceAttributes retrieves the UID string of a service and parses
+// the hostname into name and namespace. It was likely the most inefficient
+// method in Istio (and probably beyond).
+func (ps *PushStatus) GetServiceAttributes(hostname Hostname) (*ServiceAttributes, error) {
+	// The original code was calling GetService(host) in each registry, and
+	// GetServiceAttributes in the registry having the entry.
+
+	// For consul and most others, it was returning hostname, default, nil
+	// For ServiceEntryStore, returns the same thing - but checks if
+	// service exists, and returns the correct namespace (no UID) - but that's
+	// the namespace where the ServiceEntry was defined, not where
+	// the workload is running
+
+	// For k8s - it just parses the name and returns UID, while checking
+	// for existence
+
+	// TODO: lookup the service by hostname (using a hashmap), and use labels
+	// for namespace.
+
+	s, found := ps.ServiceByHostname[hostname]
+	if found && len(s.Namespace) > 0 {
+		// Namespace explicitly set. For consistency with non-k8s platform, the name
+		// will be the full name of the service.
+
+		return &ServiceAttributes{
+			Name:      string(hostname),
+			Namespace: s.Namespace,
+			UID:       fmt.Sprintf("istio://%s/services/%s", s.Namespace, hostname),
+		}, nil
+	}
+
+	// TODO: handle custom suffix
+	if strings.HasSuffix(string(hostname), ".cluster.local") {
```

Comment:
No hard-coding in `cluster.local` anymore please. It's already difficult to fish for all these TODOs.